profile/profile.go

```diff
@@ -435,6 +438,74 @@ func (p *Profile) Aggregate(inlineFrame, function, filename, linenumber, address
 	return p.CheckValid()
 }
 
+// NumLabelUnits returns a map of numeric label keys to the units
+// associated with those keys and a map of those keys to any units
+// that were encountered but not used.
+// Unit for a given key is the first encountered unit for that key. If multiple
+// units are encountered for values paired with a particular key, then the first
+// unit encountered is used and all other units are returned in sorted order
+// in map of ignored units.
+// If no units are encountered for a particular key, the unit is then inferred
+// based on the key.
+func (p *Profile) NumLabelUnits() (map[string]string, map[string][]string) {
+	numLabelUnits := map[string]string{}
+	ignoredUnits := map[string]map[string]bool{}
+
+	// Determine units based on numeric tags for each sample.
+	for _, s := range p.Sample {
+		for k, vs := range s.NumLabel {
+			units := s.NumUnit[k]
+			if len(units) != len(vs) {
+				if _, ok := numLabelUnits[k]; !ok {
+					numLabelUnits[k] = ""
+				} else {
+					ignoredUnits[k] = map[string]bool{"": true}
+				}
+				continue
+			}
+			for _, unit := range units {
+				if wantUnit, ok := numLabelUnits[k]; !ok {
+					numLabelUnits[k] = unit
+				} else if wantUnit != unit {
+					if v, ok := ignoredUnits[k]; ok {
+						v[unit] = true
+					} else {
+						ignoredUnits[k] = map[string]bool{unit: true}
+					}
+				}
+			}
+		}
+	}
+
+	// Infer units for keys without any units associated with
+	// numeric tag values.
+	for key := range numLabelUnits {
```

Comment:
This can be "for key, unit := range numLabelUnits {" I think and you don't need the lookup by key in the next condition line.